helix-core/src/test/java/org/apache/helix/integration/TestZkCallbackHandlerLeak.java

```diff
@@ -317,6 +324,150 @@ public boolean verify() throws Exception {
 
     System.out.println("END " + clusterName + " at " + new Date(System.currentTimeMillis()));
   }
+  @Test
+  public void testDanglingCallbackHanlderFix() throws Exception {
+    String className = TestHelper.getTestClassName();
+    String methodName = TestHelper.getTestMethodName();
+    String clusterName = className + "_" + methodName;
+    final int n = 3;
+    final String zkAddr = ZK_ADDR;
+    System.out.println("START " + clusterName + " at " + new Date(System.currentTimeMillis()));
+
+    TestHelper.setupCluster(clusterName, zkAddr, 12918, "localhost", "TestDB", 1, // resource
+        32, // partitions
+        n, // nodes
+        2, // replicas
+        "MasterSlave", true);
+
+    final ClusterControllerManager controller =
+        new ClusterControllerManager(zkAddr, clusterName, "controller_0");
+    controller.syncStart();
+
+    MockParticipantManager[] participants = new MockParticipantManager[n];
+    for (int i = 0; i < n; i++) {
+      String instanceName = "localhost_" + (12918 + i);
+      participants[i] = new MockParticipantManager(zkAddr, clusterName, instanceName);
+      participants[i].syncStart();
+    }
+
+    Boolean result =
+        ClusterStateVerifier
+            .verifyByZkCallback(new ClusterStateVerifier.BestPossAndExtViewZkVerifier(zkAddr,
+                clusterName));
+    Assert.assertTrue(result);
+
+    //HelixManager rpManager  = HelixManagerFactory
+    //    .getZKHelixManager(clusterName, "", InstanceType.SPECTATOR, ZK_ADDR);
+    //rpManager.connect();
+    ClusterSpectatorManager rpManager  = new ClusterSpectatorManager(ZK_ADDR, clusterName, "router");
+    rpManager.syncStart();
+    RoutingTableProvider rp = new RoutingTableProvider(rpManager, PropertyType.CURRENTSTATES);
+
+    Thread.sleep(5000);
```

Comment:
We agreed on not adding more magic numbers as the sleep time before.
Please use wait or verifier which loop and wait until the expected condition matches.

Reply:
This is very good question. Ideally we should not use sleep(). But verify some state of cluster etc. 

But in this case of testing leaking watcher etc, all these verification does not guarantee you anything. Last time, I specifically wrote an doc explaining the reason why it is like this. The example is testRemoveUserCbHandlerOnPathRemoval().

Reply:
@jiajunwang, basically I think this is a special case for watcher. If you don't think so, please give a proposal.

Reply:
" all these verification does not guarantee you anything" What kind of verification did you mean? You meant verifier? But how do you guarantee sleeping 5 secs would be 100% correct: it would return expected result?

Reply:
@kaisun2000 unverifiable case does exist. But I'm not sure if here is the same case. I'm not reading the test code yet. Could you please clarify what it is sleeping for? Then I think I can give you some ideas.

For example the later sleep, it seems you are waiting for new session established. In this case, instead of sleeping, you can use the verifier to polling until the session ID changed. Then you don't need to sleep for a hardcoded period.

Reply:
It doesn't have to explicitly sleep to wait for the new session established because expireSession() already guarantees new session established. I guess next sleep may be waiting for session handling or callback handling? But is it necessary? If necessary, could we find a better/deterministic way to verify it? Could you help understand what each of these sleeps is for.

Reply:
@jiajunwang. This is a good one for discussion as how best can we make the test more stable. 

In this sleep(), we wait for RP to get to a stable state of installing the watches, callbackhandlers. However, previous verifyWithBestPossible whatever is waiting for cluster to be stable. That is for controller. As far I know there is not equivalent for RP. And is there a way we can do similar thing for RP to reach a stable state?

line 374 sleep(). We need to wait till RP's zkHelixManager finish handleAllEvent(). By that time, it would have duplicated CB. 

line 389, we need to wait till participant session expiration finishing handleNewSession(), and RP's leaked version of CB's handleChildrenChange() finishing processing. This reproduce the dangling CB's leak on server.

Reply:
My suggestion is as follows,

Sleep one: define anonymous verifier which compares the RP snapshot with the resource idealstates. When they match, the RP get the stable state.

Sleep two: I guess you can verify if the session Id in the HelixManager has been updated or not. That is only done after the new session handling starts. Even the wait ends before the new session handling finishes, it won't do anything else before it's done anyway.

Sleep three: what if you don't wait here, but polling to verify if the childListeners are reduced to the expected number within a certain period?

Reply:
> "Sleep two: I guess you can verify if the session Id in the HelixManager has been updated or not. That is only done after the new session handling starts. Even the wait ends before the new session handling finishes, it won't do anything else before it's done anyway."

This is not going to work. Dangling CB is created in the following lambda/event created by fireAllEvent().

Reply:
Put a @todo here.

Reply:
@pkuwm would propose a framework to implement more reliable test.